program/src/processor/sync_sol_stake.rs

```diff
@@ -0,0 +1,162 @@
+        &sol_staker_stake.sol_stake,
+        ctx.accounts.config.key,
+        program_id,
+    );
+
+    require!(
+        ctx.accounts.sol_staker_stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "sol staker stake",
+    );
+
+    // validator stake
+    // - owner must be the stake program
+    // - must have the correct derivation (validates both the validator vote
+    //   and config accounts)
+    // - must be initialized
+
+    require!(
+        ctx.accounts.validator_stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "validator stake"
+    );
+
+    // validator vote must match the SOL staker stake state account's validator vote
+    // (validation done on the derivation of the expected address)
+    let (derivation, _) = find_validator_stake_pda(
+        &sol_staker_stake.delegation.validator_vote,
+        ctx.accounts.config.key,
+        program_id,
+    );
+
+    require!(
+        ctx.accounts.validator_stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "validator stake",
+    );
+
+    let mut stake_data = ctx.accounts.validator_stake.try_borrow_mut_data()?;
+    let validator_stake = unpack_initialized_mut::<ValidatorStake>(&mut stake_data)?;
+
+    // stake state (validated by the SOL Stake View program)
+    // - must match the one on the SOL staker stake account
+
+    require!(
+        ctx.accounts.sol_stake.key == &sol_staker_stake.sol_stake,
+        StakeError::IncorrectSolStakeAccount,
+        "sol stake"
+    );
+
+    require!(
+        ctx.accounts.sol_stake_view_program.key == &paladin_sol_stake_view_program_client::ID,
+        ProgramError::IncorrectProgramId,
+        "invalid sol stake view program"
+    );
+
+    GetStakeActivatingAndDeactivatingCpiBuilder::new(ctx.accounts.sol_stake_view_program)
+        .stake(ctx.accounts.sol_stake)
+        .stake_history(ctx.accounts.sysvar_stake_history)
+        .invoke()?;
+
+    let (_, return_data) = get_return_data().ok_or(ProgramError::InvalidAccountData)?;
+    let stake_state_data =
+        bytemuck::try_from_bytes::<GetStakeActivatingAndDeactivatingReturnData>(&return_data)
+            .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    // Updates the SOL stake on both the validator and SOL staker stake accounts.
+    //
+    //   1. Substract the previous stake amount from the validator stake account.
+    //
+    //   2. Updates the new stake amount on the SOL staker stake account.
+    //
+    //   3. Add the new stake amount to the validator stake account.
+
+    validator_stake.total_staked_lamports_amount = validator_stake
+        .total_staked_lamports_amount
+        .checked_sub(sol_staker_stake.lamports_amount)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+
+    sol_staker_stake.lamports_amount = u64::from(stake_state_data.activating)
+        .checked_add(stake_state_data.effective.into())
+        .ok_or(ProgramError::ArithmeticOverflow)?;
```

Comment:
If we're including `activating` in the `lamports_amount`, then we should subtract `deactivating`. I think we'll need to do this, because if we don't sync when it's deactivating, we might miss it totally if the user immediately activates onto another validator after the deactivation is complete.

Reply:
Cool, I will update to subtract `deactivating`.

Reply:
I wonder if we need to improve the logic further. Is there a possibility that we miss to sync while the stake is deactivating (e.g., the off-chain component does not run) and the stake gets activated with another validator? The "pointer" to the sol stake would still be valid, since it would be equal to the value stored on the account, so this will pass:
```rust
require!(
    ctx.accounts.sol_stake.key == &sol_staker_stake.sol_stake,
    StakeError::IncorrectSolStakeAccount,
    "sol stake"
);
```
But if we add a test for the `delegated_vote` it will missmatch since the account is delegated to a different validator.

So perhaps if we encounter this scenario we need to update the delegation to `0`.

Reply:
Improved the sync logic + added a test to simulate this scenario:
* [Improve sync logic](https://github.com/paladin-bladesmith/stake-program/pull/46/commits/b6f917d2251d192e252f9a09c872d11226c8305f)